src/governance-validators.cpp

```diff
@@ -0,0 +1,349 @@
+    : strData(),
+      objJSON(UniValue::VOBJ),
+      fJSONValid(false),
+      strErrorMessages()
+{
+    if(!strDataHexIn.empty()) {
+        SetHexData(strDataHexIn);
+    }
+}
+
+void CProposalValidator::Clear()
+{
+    strData = std::string();
+    objJSON = UniValue(UniValue::VOBJ);
+    fJSONValid = false;
+    strErrorMessages = std::string();
+}
+
+void CProposalValidator::SetHexData(const std::string& strDataHexIn)
+{
+    std::vector<unsigned char> v = ParseHex(strDataHexIn);
+    strData = std::string(v.begin(), v.end());
+    ParseJSONData();
+}
+
+bool CProposalValidator::Validate()
+{
+    if(!ValidateJSON()) {
+        return false;
+    }
+    if(!ValidateName()) {
+        strErrorMessages += "Invalid name;";
+        return false;
+    }
+    if(!ValidateStartEndEpoch()) {
+        strErrorMessages += "Invalid start:end range;";
+        return false;
+    }
+    if(!ValidatePaymentAmount()) {
+        strErrorMessages += "Invalid payment amount;";
+        return false;
+    }
+    if(!ValidatePaymentAddress()) {
+        strErrorMessages += "Invalid payment address;";
+        return false;
+    }
+    if(!ValidateURL()) {
+        strErrorMessages += "Invalid URL;";
+        return false;
+    }
+    return true;
+}
+
+bool CProposalValidator::ValidateJSON()
+{
+    return fJSONValid;
+}
+
+bool CProposalValidator::ValidateName()
+{
+    std::string strName;
+    if(!GetDataValue("name", strName)) {
+        strErrorMessages += "name field not found;";
+        return false;
+    }
+
+    std::string strNameStripped = StripWhitespace(strName);
+
+    if(strNameStripped.empty()) {
+        strErrorMessages += "name is empty;";
+        return false;
+    }
+
+    static const std::string strAllowedChars = "-_abcdefghijklmnopqrstuvwxyz012345789";
+
+    std::transform(strName.begin(), strName.end(), strName.begin(), ::tolower);
+
+    if(strName.find_first_not_of(strAllowedChars) != std::string::npos) {
+        strErrorMessages += "name contains invalid characters;";
+        return false;
+    }
+
+    return true;
+}
+
+bool CProposalValidator::ValidateStartEndEpoch()
+{
+    int64_t nStartEpoch = 0;
+    int64_t nEndEpoch = 0;
+
+    if(!GetDataValue("start_epoch", nStartEpoch)) {
+        strErrorMessages += "start_epoch field not found;";
+        return false;
+    }
+
+    if(!GetDataValue("end_epoch", nEndEpoch)) {
+        strErrorMessages += "end_epoch field not found;";
+        return false;
+    }
+
+    if(nEndEpoch <= nStartEpoch) {
+        strErrorMessages += "end_epoch <= start_epoch field not found;";
+        return false;
+    }
+
+    return true;
+}
+
+bool CProposalValidator::ValidatePaymentAmount()
+{
+    double dValue = 0.0;
+
+    if(!GetDataValue("payment_amount", dValue)) {
+        strErrorMessages += "payment_amount field not found;";
+        return false;
+    }
+
+    if(dValue <= 0.0) {
+        strErrorMessages += "payment_amount invalid;";
```

Comment:
My goal was to replicate the sentinel Proposal.is_valid() method.

However though that method doesn't check for too large amounts, such amounts will be rejected at superblock trigger creation time so I guess it would make sense to add that check here.